File: bakamc-folia/src/main/kotlin/cn/bakamc/folia/config/FlightEnergyConfig.kt
package cn.bakamc.folia.config

import moe.forpleuvoir.nebula.config.container.ConfigContainerImpl
import moe.forpleuvoir.nebula.config.item.impl.*
import net.kyori.adventure.bossbar.BossBar
import kotlin.time.Duration
import kotlin.time.Duration.Companion.minutes
import kotlin.time.Duration.Companion.seconds

object FlightEnergyConfig : ConfigContainerImpl("flight_energy") {

    val TICK_PERIOD by duration("tick_period", 1.0.seconds)

    val ENERGY_COST by double("energy_cost", 1.0)

    val MAX_ENERGY by double("max_energy", 5000.0)

    val SYNC_PERIOD by duration("sync_period", 5.0.minutes)

    val ALLOW_FLY_WORLD by stringList("allow_fly_world", listOf("world"))

    val FORBID_FLY_WORLD_MESSAGE by string("forbid_fly_world_message", "&{#FF0000}所在的世界禁止飞行")

    val CLOSE_ADVENTURE_PLAYERS_FLYING by boolean("close_adventure_players_flying", false)

    val ENERGY_PRICE_MAP by stringDoubleMap(
        "energy_price", mapOf(
            "coin_1" to 1.0
        )
    )

    val ONLINE_DURATION_DISCOUNT_MAP by stringDoubleMap(
        "online_duration_discount", mapOf(
            "10h" to 0.95,
            "1d" to 0.90,
            "7d" to 0.7,
            "30d" to 0.6,
        )
    )

    fun onlineDiscount(onlineDuration: Duration): Double {
        buildMap {
            ONLINE_DURATION_DISCOUNT_MAP.forEach { (key, value) ->
                runCatching {
                    this[Duration.parse(key)] = value.coerceIn(0.0, 1.0)
                }
            }
        }.toSortedMap { o1, o2 ->
            o2.compareTo(o1)
        }.forEach { (duration, value) ->
            if (onlineDuration >= duration) {
                return value
            }
        }
        return 1.0
    }

    val MONEY_ITEM by stringDoubleMap(
        "money_item",
        mapOf(
            "⑨币" to 5000.0,
            "冰辉石" to 78.125
        )
    )

    private val energyBar = addConfig(EnergyBar)

    object EnergyBar : ConfigContainerImpl("energy_bar") {

        val COLOR: BossBar.Color by enum("color", BossBar.Color.GREEN)

        val TITLE by string("title", "飞行能量: %.2f(%+.2f)/%.2f")

        val STYLE: BossBar.Overlay by enum("style", BossBar.Overlay.NOTCHED_10)

    }

}

File: bakamc-folia/src/main/scala/cn/bakamc/folia/config/FlightEnergyConfig.scala
package cn.bakamc.folia.config

import cn.bakamc.common.inlinestyletext.InlineStyleTextParser
import moe.forpleuvoir.nebula.config.item.*
import moe.forpleuvoir.nebula.config.{Comment, CommentLines, ConfigGroup}
import net.kyori.adventure.bossbar.BossBar
import net.kyori.adventure.text.Component

import scala.concurrent.duration.{Duration, DurationInt, FiniteDuration}
import scala.language.implicitConversions

object FlightEnergyConfig extends ConfigGroup("flight_energy") {

  @Comment("数据库同步周期,每隔一段时间将数据同步至数据库")
  val syncPeriod = ConfigFiniteDuration("sync_period", 5.minute)

  @Comment("飞行能量计算周期")
  val tickPeriod = ConfigFiniteDuration("tick_period", 1.second)

  @Comment("每次计算时消耗的飞行能量基数")
  val energyCost = ConfigDouble("energy_cost", 1)

  @Comment("玩家可以拥有的最大飞行能量")
  val maxEnergy = ConfigDouble("max_energy", 5000)

  @Comment("允许飞行的世界")
  val allowFlyWorld = ConfigList[String]("allow_fly_world", List("world"))

  @Comment("禁止飞行世界提示")
  private val forbidFlyWorldMessage = ConfigString("forbid_fly_world_message", "&{#FF0000}所在的世界禁止飞行")

  private var _forbidFlyWorldMessageComponent: Component = InlineStyleTextParser.DEFAULT.parse(forbidFlyWorldMessage)

  def forbidFlyWorldMessageComponent: Component = _forbidFlyWorldMessageComponent

  forbidFlyWorldMessage.observe { e =>
    _forbidFlyWorldMessageComponent = InlineStyleTextParser.DEFAULT.parse(e)
  }

  @Comment("是否禁止冒险模式下的飞行")
  val disableAdventureModeFly = ConfigBoolean("disable_adventure_mode_fly", false)

  @CommentLines(Array(
    "飞行能量价格映射,每1货币对应的飞行能量",
    "如果经济插件不支持Vault的多货币将会使用#default的值,#default未设置时默认为1"
  ))
  val energyPriceMap = ConfigMap[Double]("energy_price_map", Map(
    "#default" -> 1.0,
    "coin_1" -> 1.0
  ))

  def currencies: List[String] = energyPriceMap.keys.map(s => s"\"$s\"").toList

  def energyPrice(currency: String): BigDecimal =
    BigDecimal.valueOf(energyPriceMap.get(currency).getOrElse(energyPriceMap.get("#default").getOrElse(1.0)))

  @Comment("在线时长折扣映射,在线时长达到对应时间时获得相应的折扣")
  private val _onlineTimeDiscountMap = ConfigMap[Double]("online_time_discount_map", Map(
    10.hour.toString -> 0.95,
    1.day.toString -> 0.9,
    7.day.toString -> 0.7,
    30.day.toString -> 0.6
  ))

  def onlineTimeDiscountMap: Map[FiniteDuration, Double] = _onlineTimeDiscountMap.getValue.map {
    case (str, d) => (Duration(str).asInstanceOf[FiniteDuration], d.max(0.1).min(1.0))
  }

  def onlineDiscount(onlineDuration: FiniteDuration): Double = {
    val discountMap = onlineTimeDiscountMap
    discountMap.keys.foldLeft(Option.empty[FiniteDuration]) {
      case (None, d) if d < onlineDuration => Some(d)
      case (Some(max), d) if d < onlineDuration && d > max => Some(d)
      case (acc, _) => acc
    }.map(discountMap.apply).getOrElse(1.0)
  }

  @Comment("货币物品映射,物品名称对应物品价格")
  val moneyItem = ConfigMap[Double]("moneyItem", Map(
    "⑨币" -> 5000,
    "冰辉石" -> 78.125
  ))

  object EnergyBar extends ConfigGroup("energy_bar") {

    @Comment("能量条颜色[PINK ,BLUE ,RED ,GREEN ,YELLOW ,PURPLE, WHITE]")
    val color = ConfigJavaEnum[BossBar.Color]("color", BossBar.Color.GREEN)

    @Comment("能量条标题")
    val title = ConfigString("title", "飞行能量: %.2f(%+.2f)/%.2f")

    @Comment("能量条样式[PROGRESS, NOTCHED6, NOTCHED10, NOTCHED12, NOTCHED20]")
    val style = ConfigJavaEnum[BossBar.Overlay]("style", BossBar.Overlay.NOTCHED_10)

  }

}
